### PR TITLE
feat(): npm star

### PR DESCRIPTION
**Pull request by Izak88**

### CODE REVIEW

File: src/api/auth.ts [hunk 1]
```diff
@@ -407,7 +407,8 @@ export function publishPackage(
                 owners: [ username ],
                 organization: organization ? organization : '',
                 teamPermissions: permissions,
-                memberPermissions: [{ name: username, permission: 'read-write' }]
+                memberPermissions: [{ name: username, permission: 'read-write' }],
```
Comment (jkuri) [inline src/api/auth.ts]: what do you think of format for permissions `read: true, write: false`?

File: src/api/routes.ts [hunk 2]
```diff
@@ -224,43 +224,71 @@ export function getTarball(req: auth.AuthRequest, res: express.Response) {
...
-    let splitName = name.split('/')[0].slice(1);
-    if (splitName !== user.name) {
-      organization = splitName;
-      teams = auth.getUserTeams(user.name, organization, authFile).map(t => {
-        return { team: t.name, permission: 'read-write' };
+  let request = req.headers.referer.split(' ');
+  if (request[0] === 'star') {
+    let username = user.name;
+    if (username in req.body.users) {
+      auth.starPackage(user.name, name, authFile).then((newAuthFile) => {
```
Comment (jkuri) [inline src/api/routes.ts:234]: why is `starPackage` called inside `publishPackage` function? is this related?

Comment (Izak88) [reply src/api/routes.ts:234]: I see what you mean... Well yeah, the same route is called when you want to publish a package or star it... I think I should rename this function into something similar like updatePackage...

File: src/api/auth.ts (hunk 1 above)
Comment (Izak88) [reply src/api/auth.ts]: Actually, I don't like either :), but yeah I think your's version is a little nicer... I'll change into that format....

Comment (jkuri) [reply src/api/auth.ts]: what format do you think would be okay and cleaner? open to suggestions.

Comment (Izak88) [reply src/api/auth.ts]: I don't know, maybe we could use only `write: false / true`, because if the user will have a record here, he'll automatically have at least read permissions, so read will always be true... But on the other hand, if we put `read: true` it will be more intuitive so I think we could use your's format... What do you think?

Comment (jkuri) [reply src/api/auth.ts]: your objection convinced me :) there's no reason to have `read: true` if its always the same value.  maybe just `permissions: true / false` in that case?

Comment (irmana) [reply src/api/auth.ts]: @Izak88 @jkuri consider the use case when admin wants to temporary restrict the access to some packages or hide them from certain users, then you just i.e. set `read: false`. In case of Jan's format you don't need to parse anything.

Comment (Izak88) [reply src/api/auth.ts]: Yeah, Jan's format is definitely more open and leaves us with open options in the future...

File: src/api/routes.ts (hunk 2 above)
Comment (Izak88) [reply src/api/routes.ts:234]: done

File: src/api/auth.ts (hunk 1 above)
Comment (Izak88) [reply src/api/auth.ts]: done

File: src/api/routes.ts [hunk 3]
```diff
@@ -223,44 +223,72 @@ export function getTarball(req: auth.AuthRequest, res: express.Response) {
...
       }
-  }).catch(() => res.status(412).json(jsonErrorResponse) );
+    }
+    let data: IPackage = req.body;
+    let version: string = data.versions[Object.keys(data.versions)[0]].version;
+
+    auth.publishPackage(
+      name, user.name, organization, teams, version, authFile).then((newAuthFile) => {
+        if (!newAuthFile) {
+          return res.status(403).json({ error: `you do not have permission to publish `
```
Comment (irmana) [inline src/api/routes.ts:275]: Is this text supposed to be here or in the frontend lang file?

File: src/api/routes.ts [hunk 4]
```diff
@@ -223,44 +223,72 @@ export function getTarball(req: auth.AuthRequest, res: express.Response) {
...
+              + `"${name}". Are you logged in as the correct user?` });
+        } else {
+          writeJsonFile(getAuthPath(), newAuthFile).then(() => {
+            let pkg = new Package(data);
+            pkg.saveTarballFromData()
+              .then(() => pkg.initPkgJsonFromData())
+              .then(() => {
+                return res.status(200).json({ message: 'package published' });
```
Comment (irmana) [inline src/api/routes.ts:283]: same here, hardcoded text msg...

File: src/api/routes.ts (hunk 3 above)
Comment (Izak88) [reply src/api/routes.ts:275]: a string has to be returned here because this function is called from npm and then prints the error message, but we can generate lang file with all the strings and use it here... Did you have that in mind?

**Review by jkuri — APPROVED** [review]
I believe this can go in.
